### PR TITLE
feat: remove SOCKS5 roundtrip

### PR DESCRIPTION
**Pull request by fortuna**
This PR is a rewrite of the SOCKS5 code to no longer use x/proxy.
Now we can sent the method and the connect requests in one write, avoiding a roundtrip.

Later, our rewrite will allow us to merge initial client data with the connection, making the initial packet size less predictable, like we do for Shadowsocks: https://github.com/Jigsaw-Code/outline-sdk/blob/ab1e88d4a2f9057c7096dec360f477086b9a62ef/transport/shadowsocks/stream_dialer.go#L93

We still need to figure out how to properly handle this kind of merging in a way that is reusable across dialers. Currently the Shadowsocks dialer has its own implementation.

### CODE REVIEW

**Review by fortuna — COMMENTED** [review]
Any thoughts?

File: transport/socks5/socks5.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+	"net"
+	"strconv"
+)
+
+const addrTypeIPv4 = 0x01
+const addrTypeDomainName = 0x03
+const addrTypeIPv6 = 0x04
+
+func appendSOCKS5Address(b []byte, address string) ([]byte, error) {
```
Comment (jyyi1) [inline transport/socks5/socks5.go:28]: *nit* maybe it's worth to put a reference link or just a bytes diagram of the related SOCK5 protocol to this function?

File: transport/socks5/stream_dialer.go [hunk 2]
```diff
@@ -18,10 +18,9 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"net"
+	"io"
 
 	"github.com/Jigsaw-Code/outline-sdk/transport"
-	"golang.org/x/net/proxy"
```
Comment (jyyi1) [inline transport/socks5/stream_dialer.go:24]: run `go mod tidy` to remove the package from `go.mod`?

**Review by jyyi1 — APPROVED** [review]
LGTM, also I think it will be helpful to mention in the comment that why we are not using `x/net/proxy`.

File: transport/socks5/socks5.go (hunk 1 above)
Comment (fortuna) [reply transport/socks5/socks5.go:36]: Good idea. Added.

File: transport/socks5/stream_dialer.go (hunk 2 above)
Comment (fortuna) [reply transport/socks5/stream_dialer.go:24]: Done! I forgot about it.

**Pull request merged**